### PR TITLE
[Lake] Impl publish version for write task

### PR DESCRIPTION
**Pull request by sduzh**
## What type of PR is this：
- [ ] bug
- [ ] feature
- [ ] enhancement
- [ ] refactor
- [x] others

### CODE REVIEW

**Review by imay — DISMISSED** [review]

File: be/src/service/service_be/lake_service.cpp [hunk 1]
```diff
@@ -4,27 +4,146 @@
...
+    // TODO move the execution to TaskWorkerPool
+    for (const auto& tablet_id : request->tablet_ids()) {
+        auto tablet = _env->lake_tablet_manager()->get_tablet(tablet_id);
+        if (!tablet.ok()) {
+            LOG(WARNING) << "Fail to get tablet " << tablet_id << ": " << tablet.status();
+            response->add_failed_tablets(tablet_id);
+            continue;
+        }
+        auto base_metadata = tablet->get_metadata(base_version);
+        if (base_metadata.status().is_not_found() && tablet->get_metadata(request->version()).ok()) {
```
Comment (wyb) [inline be/src/service/service_be/lake_service.cpp:77]: Why check base tablet metadata `is_not_found`?

File: be/src/service/service_be/lake_service.cpp [hunk 2]
```diff
@@ -4,27 +4,146 @@
...
+            response->add_failed_tablets(tablet_id);
+            continue;
+        }
+
+        // make a copy of metadata
+        auto new_metadata = std::make_shared<lake::TabletMetadata>(**base_metadata);
+        new_metadata->set_version(request->version());
+
+        bool succeeded = false;
+        for (auto txn_id : request->txn_ids()) {
```
Comment (wyb) [inline be/src/service/service_be/lake_service.cpp:91]: Does the tablet metadata between base version and publish version need to be generated?

File: be/src/service/service_be/lake_service.cpp [hunk 3]
```diff
@@ -4,27 +4,146 @@
...
+                LOG(WARNING) << "Fail to apply " << tablet->txn_log_path(txn_id) << ": " << st;
+                response->add_failed_tablets(tablet_id);
+                break;
+            }
+            succeeded = true;
+        }
+        if (!succeeded) {
+            continue;
+        }
+        auto st = tablet->put_metadata(new_metadata);
```
Comment (wyb) [inline be/src/service/service_be/lake_service.cpp:113]: The case below may happen.
Only part of the txns apply successful and delete all txn logs at the end.

File: be/src/service/service_be/lake_service.cpp (hunk 2 above)
Comment (sduzh) [reply be/src/service/service_be/lake_service.cpp]: No

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]


:heart_eyes: **pass** : 0 / 0 (0%)

File: be/src/service/service_be/lake_service.cpp (hunk 1 above)
Comment (sduzh) [reply be/src/service/service_be/lake_service.cpp]: The GC module may have deleted the old version metadata.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]


:heart_eyes: **pass** : 0 / 0 (0%)

File: be/src/service/service_be/lake_service.cpp (hunk 2 above)
Comment (sduzh) [reply be/src/service/service_be/lake_service.cpp]: There is no need, at least now,  to generate multiple versions in a single publish version request.

File: be/src/service/service_be/lake_service.cpp (hunk 3 above)
Comment (sduzh) [reply be/src/service/service_be/lake_service.cpp]: Sorry, I didn't grasp your question.

**Comment by sduzh** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]


:heart_eyes: **pass** : 0 / 0 (0%)

File: be/src/service/service_be/lake_service.cpp (hunk 2 above)
Comment (wyb) [reply be/src/service/service_be/lake_service.cpp]: Do not consider time travel?

File: gensrc/proto/starlake.proto [hunk 4]
```diff
@@ -63,8 +63,8 @@ message TxnLogPB {
 message PublishVersionRequest {
     repeated int64 tablet_ids = 1;
     repeated int64 txn_ids = 2;
-    // The version to be published. base version would be version - txn_ids_size()
-    optional int64 version = 3;
+    optional int64 base_version = 3;
```
Comment (wyb) [inline gensrc/proto/starlake.proto:66]: Why did you add `base_version` back?

File: be/src/service/service_be/lake_service.cpp (hunk 3 above)
Comment (sduzh) [reply be/src/service/service_be/lake_service.cpp]: This won't happen in the current implementation.

File: gensrc/proto/starlake.proto (hunk 4 above)
Comment (sduzh) [reply gensrc/proto/starlake.proto:66]: This is no relationship between the base version and the size of txn_ids if one of the transactions is a schema change transaction.

File: be/src/service/service_be/lake_service.cpp (hunk 1 above)
Comment (sduzh) [reply be/src/service/service_be/lake_service.cpp]: comment updated, PTAL

File: gensrc/proto/starlake.proto (hunk 4 above)
Comment (sduzh) [reply gensrc/proto/starlake.proto:68]: Add some comments, PTAL

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Review by wyb — DISMISSED** [review]

File: be/src/service/service_be/lake_service.cpp [hunk 5]
```diff
@@ -4,27 +4,159 @@
...
-    (void)request;
-    response->mutable_status()->set_status_code(TStatusCode::NOT_IMPLEMENTED_ERROR);
+
+    // TODO: move the execution to TaskWorkerPool
+    // This rpc never fail.
+    response->mutable_status()->set_status_code(TStatusCode::OK);
+    for (const auto& tablet_id : request->tablet_ids()) {
+        auto tablet = _env->lake_tablet_manager()->get_tablet(tablet_id);
+        if (!tablet.ok()) {
+            LOG(WARNING) << tablet.status();
```
Comment (wuleistarrocks) [inline be/src/service/service_be/lake_service.cpp:151]: more detailed warning message?

File: be/src/service/service_be/lake_service.cpp [hunk 6]
```diff
@@ -4,27 +4,159 @@
...
 
 namespace starrocks {
 
+inline Status apply_txn_log(const lake::TxnLog& log, lake::TabletMetadata* metadata) {
+    if (log.has_op_write()) {
+        if (log.op_write().has_rowset() && log.op_write().rowset().segments_size() > 0) {
+            auto rowset = metadata->add_rowsets();
+            rowset->CopyFrom(log.op_write().rowset());
+            rowset->set_id(metadata->next_rowset_id());
+            metadata->set_next_rowset_id(metadata->next_rowset_id() + std::max<int>(1, rowset->segments_size()));
```
[the remark this answers is not in the document]
Comment (sduzh) [reply be/src/service/service_be/lake_service.cpp:22]: rowset->segments_size()  can be zero.

File: be/src/service/service_be/lake_service.cpp (hunk 5 above)
Comment (sduzh) [reply be/src/service/service_be/lake_service.cpp]: updated

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Review by wuleistarrocks — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Review by wuleistarrocks — APPROVED** [review]

**Review by wyb — APPROVED** [review]

**Review by chaoyli — APPROVED** [review]
